cadquery/occ_impl/solver.py

```diff
@@ -20,36 +102,317 @@
+
+        return rv.toDir()
+
+    def _getPln(self, arg: Shape) -> gp_Pln:
+
+        if isinstance(arg, Face):
+            rv = gp_Pln(self._getPnt(arg), arg.normalAt().toDir())
+        elif isinstance(arg, (Edge, Wire)):
+            normal = arg.normal()
+            origin = arg.Center()
+            plane = Plane(origin, normal=normal)
+            rv = plane.toPln()
+        else:
+            raise ValueError(f"Can not construct a plane for {arg}.")
+
+        return rv
+
+    def _getPnt(self, arg: Shape) -> gp_Pnt:
+
+        # check for infinite face
+        if isinstance(arg, Face) and any(
+            Precision.IsInfinite_s(x) for x in BRepTools.UVBounds_s(arg.wrapped)
+        ):
+            # fall back to gp_Pln center
+            pln = arg.toPln()
+            center = Vector(pln.Location())
+        else:
+            center = arg.Center()
+
+        return center.toPnt()
+
+    def _getLin(self, arg: Shape) -> gp_Lin:
+
+        if isinstance(arg, (Edge, Wire)):
+            center = arg.Center()
+            tangent = arg.tangentAt()
+        else:
+            raise ValueError(f"Can not construct a plane for {arg}.")
+
+        return gp_Lin(center.toPnt(), tangent.toDir())
+
+    def toPODs(self) -> Tuple[Constraint, ...]:
+        """
+        Convert the constraint to a representation used by the solver.
+
+        NB: Compound constraints are decomposed into simple ones.
+        """
+
+        # apply sublocation
+        args = tuple(
+            arg.located(loc * arg.location())
+            for arg, loc in zip(self.args, self.sublocs)
+        )
+
+        markers: List[Tuple[ConstraintMarker, ...]]
+
+        # convert to marker objects
+        if self.kind == "Axis":
+            markers = [(self._getAxis(args[0]), self._getAxis(args[1]),)]
+
+        elif self.kind == "Point":
+            markers = [(self._getPnt(args[0]), self._getPnt(args[1]))]
+
+        elif self.kind == "FixedPoint":
+            markers = [(self._getPnt(args[0]),)]
+
+        elif self.kind == "Plane":
+            markers = [
+                (self._getAxis(args[0]), self._getAxis(args[1]),),
+                (self._getPnt(args[0]), self._getPnt(args[1])),
+            ]
+
+        elif self.kind == "PointInPlane":
+            markers = [(self._getPnt(args[0]), self._getPln(args[1]))]
+
+        elif self.kind == "PointOnLine":
+            markers = [(self._getPnt(args[0]), self._getLin(args[1]))]
+
+        elif self.kind == "Fixed":
+            markers = [(None,)]
+
+        elif self.kind == "FixedPoint":
+            markers = [(self._getPnt(args[0]),)]
+
+        elif self.kind == "FixedAxis":
+            markers = [(self._getAxis(args[0]),)]
+
+        elif self.kind == "FixedRotation":
+            markers = [(None,), (None,), (None,)]
+
+        elif self.kind == "FixedRotationAxis":
+            markers = [(None,)]
+
+        else:
+            raise ValueError(f"Unknown constraint kind {self.kind}")
```

Comment:
Just incase this idea didn't occur to you when you were writing this, I think you could replace all the `if ... elif` with a dict containing lambdas, like:
```python
marker_lookup = {
    ...
    "FixedAxis": lambda s, a: [(s._getAxis(a[0]),)],
    ...
}
try:
    markers = marker_dict[self.kind](self, args)
except KeyError:
    raise ValueError(f"Unknown constraint kind {self.kind}")
```
But I can't figure out how to get that past mypy, so I'll just leave the suggestion with you.

Reply:
Yes, it'll involve some funky typing... Let's leave as-is